spring-cloud-starter-stream-source-sftp/src/main/java/org/springframework/cloud/stream/app/sftp/source/SftpSourceConfiguration.java

```diff
@@ -168,31 +199,80 @@ else if (properties.isListOnly() || properties.isTaskLauncherOutput()) {
 				.get();
 	}
 
-	private Consumer<SourcePollingChannelAdapterSpec> consumerSpec() {
-		return spec -> spec.poller(SftpSourceConfiguration.this.defaultPoller);
+	private IntegrationFlow listingFlow(SessionFactory<LsEntry> sftpSessionFactory) {
+		if (this.properties.isMultiSource()) {
+			return multiSourceListingFlow();
+		}
+		else {
+			return singleSourceListingFlow(sftpSessionFactory);
+		}
+	}
+
+	private IntegrationFlow singleSourceListingFlow(SessionFactory<LsEntry> sftpSessionFactory) {
+		return IntegrationFlows.from(() -> this.properties.getRemoteDir(), consumerSpec(this.listFilesRotator))
+				.handle(Sftp.outboundGateway(sftpSessionFactory,
+						AbstractRemoteFileOutboundGateway.Command.LS.getCommand(), "payload")
+						.options(AbstractRemoteFileOutboundGateway.Option.NAME_ONLY.getOption()))
+				.split()
+				.channel(this.properties.isListOnly() ? sftpFileListChannel() : sftpFileTaskLaunchChannel())
+				.get();
+	}
+
+	private IntegrationFlow multiSourceListingFlow() {
+		IntegrationFlowBuilder flow = IntegrationFlows.from(() ->
+					this.listFilesRotator.getCurrentDirectory(), consumerSpec(this.listFilesRotator))
+				.handle(Sftp.outboundGateway(this.delegatingSessionFactory.getFactory(),
+						AbstractRemoteFileOutboundGateway.Command.LS.getCommand(), "payload")
+						.options(AbstractRemoteFileOutboundGateway.Option.NAME_ONLY.getOption()));
+		if (this.properties.isTaskLauncherOutput()) {
+			flow.enrichHeaders(this.listFilesRotator.headers());
+		}
+		return flow
+				.handle(this.listFilesRotator, "clearKey")
```

Comment:
Is it expected do not call this in case of `singleSourceListingFlow()`?

Reply:
Correct; we don't use a delegating factory (or rotator) for a single source.

When we are multi-source, the poller sets the key in `beforeReceive()`. Unlike the `RotatingServerAdvice`, we can't clear the key in `afterReceive()` since the downstream flow (gateway) needs the key in the TL to lookup the delegate SF. After the gateway operation is complete, we clear the key.